### ISSUE TITLE
"Package or namespace load failed"

### ISSUE THREAD

**Issue by simon-tarr** [Azure/doAzureParallel#277]
Before submitting a bug please check the following:
- [ x ] Start a new R session
- [ x ] Check your credentials file
- [ x ] Install the latest doAzureParallel package
- [ x ] Submit a minimal, reproducible example 
- [ x ] run `sessionInfo()`

**Description**

I have built a new cluster using rocker/geospatial:latest and supplemented this container by installing a package from a GitHub repository at "mrke/NicheMapR". The cluster appears to start up with no errors and all nodes in the pool are idle. I am running doAzureParallel v0.7.0 and rAzureBatch v0.6.0.

I have run sessionInfo() at this point and I see the following. Note that NicheMapR isn't present.
```

R version 3.5.0 (2018-04-23)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Debian GNU/Linux 9 (stretch)

Matrix products: default
BLAS: /usr/lib/openblas-base/libblas.so.3
LAPACK: /usr/lib/libopenblasp-r0.2.19.so

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C               LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=C              LC_PAPER=en_US.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C             LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

loaded via a namespace (and not attached):
[1] compiler_3.5.0   codetools_0.2-15 iterators_1.0.9  foreach_1.4.4   
```

I then check the bottom of stderr.txt and find the following:

```
Downloading GitHub repo mrke/nichemapr@master
from URL https://api.github.com/repos/mrke/nichemapr/zipball/master
Installing NicheMapR
'/usr/local/lib/R/bin/R' --no-site-file --no-environ --no-save --no-restore  \
  --quiet CMD INSTALL  \
  '/tmp/RtmpCJ5NHz/devtools119977956/mrke-NicheMapR-2d46c4b'  \
  --library='/mnt/batch/tasks/shared/R/packages' --install-tests
* installing *source* package ‘NicheMapR’ ...
** libs
installing to /mnt/batch/tasks/shared/R/packages/NicheMapR/libs
** R
** data
*** moving datasets to lazyload DB
** inst
** byte-compile and prepare package for lazy loading
** help
*** installing help indices
** building package indices
** installing vignettes
** testing if installed package can be loaded
* DONE (NicheMapR)
```

Based on this I conclude that NicheMapR has installed correctly.

To test whether NicheMapR can load, I can the following code:

```
foreach(i = 1, .options.azure = opt, .errorhandling="pass") %dopar% {
  library(NicheMapR)
  x<-sessionInfo()
  return(x)
}
```
However the output says:

```
<simpleError: package or namespace load failed for â€˜NicheMapRâ€™:
 .onLoad failed in loadNamespace() for 'NicheMapR', details:
  call: dyn.load(micro_lib)
  error: unable to load shared object '/mnt/batch/tasks/workitems/job20180531115712/job-1/jobpreparation/wd/NicheMapR/libs/linux/MICROCLIMATE.so':
  /mnt/batch/tasks/workitems/job20180531115712/job-1/jobpreparation/wd/NicheMapR/libs/linux/MICROCLIMATE.so: cannot open shared object file: No such file or directory>

```
If I explicitly include github="mrke/NicheMapR" within the foreach loop, I get the following merge error:

```
An error has occurred in the merge task of the job 'job20180531130405'. Error handling is set to 'stop' and has proceeded to terminate the job. The user will have to handle deleting the job. If this is not the correct behavior, change the errorhandling property to 'pass'  or 'remove' in the foreach object. Use the 'getJobFile' function to obtain the logs. For more information about getting job logs, follow this link: https://github.com/Azure/doAzureParallel/blob/master/docs/40-troubleshooting.md#viewing-files-directly-from-compute-nodeError in e$fun(obj, substitute(ex), parent.frame(), e$data) : 
  object 'results' not found
```
After all that, my question is quite simple: is the problem with doAzureParallel or with NicheMapR? Any assistance would be highly appreciated. Many thanks.

**Comment by brnleehng** [comment]
Hi Simon,

I'm still investigating on why the package installation is failing. It appears R is trying to find the MICROCLIMATE.so in the job package directory. However the NicheMapR package is installed in the pool package directory. 

However, I have a PR fix for the github package installation on the job level. This will allow you to install the package at job level properly.

``` R
foreach(i = 1:3, githubPackages = c("mrke/NicheMapR")) %dopar% {
library(NicheMapR);
x <- sessionInfo();
return(x)
}
``` 

Thanks,
Brian

**Comment by brnleehng** [comment]
I think the issue with MICROCLIMATE.so being found in the wrong lib path is NicheMapR issue, https://github.com/mrke/NicheMapR/issues/4 . However, you can avoid this issue by installing it at the job level.

**Comment by simon-tarr** [comment]
Thanks Brian for the speedy fix. After reinstalling doAzureParallel to grab the fix I can confirm that my package is now successfully loading into my cluster and I can run an iteration without error.

However, since installing the latest update, I have come across a weird chunking issue which I'm not sure is related to this problem or not. I have scaled up my job from one iteration of the loop to the full run (10,670 iterations) but I'm only able to get the results of the first iteration to run.

I run:

```
cs <- ceiling(nrow(points) / getDoParWorkers())
opt <- list(chunkSize = cs)
```

over my cluster of 8 F16 nodes (128 cores total) to get a chunkSize of 85 (opt=85). When I submit this job I see the following:

```
==============================================================================================================================
Id: job20180601095335
chunkSize: 85
enableCloudCombine: TRUE
githubPackages: 
	simon-tarr/NicheMapR; 
errorHandling: pass
wait: TRUE
autoDeleteJob: TRUE
==============================================================================================================================
Submitting tasks (1/1)
Submitting merge task. . .
Job Preparation Status: Package(s) being installed..
Waiting for tasks to complete. . .
| Progress: 100.00% (1/1) | Running: 0 | Queued: 0 | Completed: 1 | Failed: 0 |
Tasks have completed. Merging results.. Completed.
```

As you can see, the chunkSize is 85 but it's only submitting one job, not 128 (I think it's 128 tasks - one for each core in my pool...is this correct?). Within cluster.json I've set maxTasksPerNode to 16 to match the F16 machines. Before this update, I had no issues with the number of tasks being sent to the cloud they would simply error because the package wasn't installed properly. Now I have the opposite problem. The package is installed fine but I can't submit the correct number of tasks to each core!

To confirm, I have installed the  latest versions of doAzureParallel and rAzureBatch and I've restarted by cluster several times (as well as my R session). 

My full code is as follows:

```
setCredentials("credentials.json")
cluster <- doAzureParallel::makeCluster("cluster.json")
registerDoAzureParallel(cluster)
getDoParWorkers()

puerto.rico<-stack("c:/globalclimate/global_climate.nc")
prico.points<-data.frame(rasterToPoints(puerto.rico[[1]]))
prico.points$X1<-NULL

cs <- ceiling(nrow(prico.points) / getDoParWorkers())
opt <- list(chunkSize = cs)

climate<-foreach(i = 1:nrow(prico.points), .options.azure = opt, .errorhandling="pass", githubPackages=c("simon-tarr/NicheMapR")) %dopar% {
  library(NicheMapR)

  micro<-NicheMapR::micro_global(loc=prico.points[i,], nyears=1, minshade = 0, maxshade = 100)
  
}
```

Is there anything obviously wrong here? Have some options been depreciated recently which is causing an issue with sending the correct number of tasks?

**Comment by simon-tarr** [comment]
Hi Brian, thanks for the speedy fix. I can confirm that I have been able to successfully load NicheMapR via this method. I'm still getting several merge errors but will need some time to view the error logs to fully understand what the problem is.

**Comment by simon-tarr** [comment]
Hi Brian, a quick update from some extra testing I've been doing.

When I first ran your code above (after updating the package), it ran fine. On subsequent testing I received errors along the lines of:

```
An error has occurred in the merge task of the job 'job20180601115003'. Error handling is set to 'stop' and has proceeded to terminate the job. The user will have to handle deleting the job. If this is not the correct behavior, change the errorhandling property to 'pass'  or 'remove' in the foreach object. Use the 'getJobFile' function to obtain the logs. For more information about getting job logs, follow this link: https://github.com/Azure/doAzureParallel/blob/master/docs/40-troubleshooting.md#viewing-files-directly-from-compute-nodeError in e$fun(obj, substitute(ex), parent.frame(), e$data) : 
  object 'results' not found
```
I found that even if I had many iterations to run (let's say 100), it would only submit one task. For example (on a single F2 machine):

```
Id: job20180601115003
chunkSize: 50
enableCloudCombine: TRUE
githubPackages: 
	mrke/NicheMapR; 
errorHandling: pass
wait: TRUE
autoDeleteJob: TRUE
==============================================================================================================================================
Submitting tasks (1/1)
Submitting merge task. . .
Job Preparation Status: Package(s) being installed.
Waiting for tasks to complete. . .
| Progress: 100.00% (1/1) | Running: 0 | Queued: 0 | Completed: 1 | Failed: 1 |
Tasks have completed. Merging results.... 
```
I found that changing the argument githubPackages to just github would then force all of the tasks to submit. However, I am still repeatedly getting the error:

`An error has occurred in the merge task of the job 'job20180601115112'. Error handling is set to 'stop' and has proceeded to terminate the job. The user will have to handle deleting the job. If this is not the correct behavior, change the errorhandling property to 'pass'  or 'remove' in the foreach object. Use the 'getJobFile' function to obtain the logs. For more information about getting job logs, follow this link: https://github.com/Azure/doAzureParallel/blob/master/docs/40-troubleshooting.md#viewing-files-directly-from-compute-nodeError in e$fun(obj, substitute(ex), parent.frame(), e$data) : 
  object 'results' not found`


Would it be possible to get some clarification as to where I need to specify packages and what the arguments are called? I was under the impression that the argument for GitHub packages was 'github' and not 'githubPackages'. Further, I'm not sure when I should be telling my cluster to install: should it be within the cluster.json file? Should it be within the foreach loop itself?

Many thanks.

**Comment by brnleehng** [comment]
@simon-tarr 

Woops, sorry for the typo. The correct argument is github, not githubPackages. 

``` R
foreach(i = 1:3, github = c("mrke/NicheMapR")) %dopar% {
library(NicheMapR);
x <- sessionInfo();
return(x) }
```
This will return 3 tasks. Using githubPackages will only create one task. 

The error is shown above, are there any logs for the error? Is it running sessionInfo()?

**Installing R packages at the cluster level** 
You should only install R packages at the cluster level when the R package are very large to install and plan on using multiple foreach loops. This will reduce your job times because you don't have to reinstall the package for every foreach loop.

**Installing R packages within the foreach loop** 
Installing R packages at the foreach level will allow you to have a clean R package directory every single time. 

For this case, you want to install NicheMapR within the foreach loop itself because I believe there's a bug in NicheMapR package that is misreading the location for MICROCLIMATE.so file. 

Thanks,
Brian

**Comment by brnleehng** [comment]
@simon-tarr 

Are you still running into the same issue with the fix?

Thanks,
Brian

**Comment by simon-tarr** [comment]
Hi @brnleehng sorry for the radio silence, I've been away from the office. I haven't had a chance yet to properly test this given the above but will do so as soon as possible and report back. Thanks again for your help.

**Comment by simon-tarr** [comment]
Hi @brnleehng a quick update on the above. 
As recommended, I have removed all package installation from the cluster.json file. I'm now calling all packages within the foreach loop like so:

`
climate<-foreach(i = 1:3, .options.azure = opt, .errorhandling="pass", github=c("mrke/nichemapr"), .packages=c("dismo")) %dopar% {
  library(NicheMapR)
  library(raster)
  library(ncdf4)
  library(dismo)
  x<-sessionInfo()
  return(x)
  
}`

In the above example the packages 'raster' and 'ncdf4' are included within rocker/geospatial:latest so I didn't call them using .packages. 'dismo' isn't included so I've called it using .packages.

The first time I ran it this morning, (upon provisioning a new cluster) it worked fine. On subsequent runs, however, I receive merge errors. e.g:

> An error has occurred in the merge task of the job 'job20180611123152'. Error handling is set to 'stop' and has proceeded to terminate the job. The user will have to handle deleting the job. If this is not the correct behavior, change the errorhandling property to 'pass'  or 'remove' in the foreach object. Use the 'getJobFile' function to obtain the logs. For more information about getting job logs, follow this link: https://github.com/Azure/doAzureParallel/blob/master/docs/40-troubleshooting.md#viewing-files-directly-from-compute-nodeError in e$fun(obj, substitute(ex), parent.frame(), e$data) : 
>   object 'results' not found

Since then I have re provisioned a new cluster but continue to get the merge error.

So far I've been unable to work out any sort of pattern as to why it works/doesn't work so I can't really provide much more information at the moment, I'm afraid.
